onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc

```diff
@@ -689,6 +693,420 @@ Status ApplyProfileShapesFromInputTensorValue(std::vector<nvinfer1::IOptimizatio
   return Status::OK();
 }
 
+/*
+ * Set TensorRT execution context input.
+ *
+ * There are two types of input tensor: (1) shape tensor and (2) execution tensor.
+ * The input buffer binding needs to be handled differently.
+ *
+ */
+Status BindContextInput(Ort::KernelContext& ctx,
+                        nvinfer1::ICudaEngine* trt_engine,
+                        nvinfer1::IExecutionContext* trt_context,
+                        const char* input_name,
+                        size_t input_index,
+                        std::vector<int32_t>& shape_values,  // only for "shape tensor"
+                        std::vector<IAllocatorUniquePtr<void>>& scratch_buffers,
+                        OrtAllocator* alloc,
+                        cudaStream_t stream) {
+  auto input_tensor = ctx.GetInput(input_index);
+  auto tensor_info = input_tensor.GetTensorTypeAndShapeInfo();
+  const auto tensor_shapes = tensor_info.GetShape();
+  const auto tensor_type = tensor_info.GetElementType();
+
+  if (trt_engine->isShapeInferenceIO(input_name)) {
+    // Get the shape value of "shape tensor"
+    if (shape_values.empty()) {
+      auto status = GetShapeOfShapeTensor(input_tensor, shape_values, trt_engine, input_name, stream);
+      if (status != Status::OK()) {
+        return ORT_MAKE_STATUS(ONNXRUNTIME, EP_FAIL, status.ErrorMessage());
+      }
+    }
+
+    // Bind "shape tensor" input buffer
+    if (!trt_context->setTensorAddress(input_name, &shape_values[0])) {
+      std::string error_input_name = input_name;
+      ORT_THROW_IF_ERROR(ORT_MAKE_STATUS(ONNXRUNTIME, EP_FAIL,
+                                         "TensorRT EP failed to call nvinfer1::IExecutionContext::setTensorAddress() for shape input '" + error_input_name + "'"));
+    }
+  } else {
+    // Set shape for input tensor which is execution tensor
+    nvinfer1::Dims dims = trt_context->getTensorShape(input_name);
+    int nb_dims = dims.nbDims;
+    for (int j = 0, end = nb_dims; j < end; ++j) {
+      dims.d[j] = static_cast<int32_t>(tensor_shapes[j]);
+    }
+    if (!trt_context->setInputShape(input_name, dims)) {
+      std::string error_input_name = input_name;
+      ORT_THROW_IF_ERROR(ORT_MAKE_STATUS(ONNXRUNTIME, EP_FAIL,
```

Comment:
Running this command: `onnxruntime_perf_test -e tensorrt -M -I -F "0:2" -F "2:256" -F "3:256" trt_path_embed.onnx` did not work for the model that is embedded like this:
```
['input1']
[<DataType.HALF: 1>]
[(-1, 8, -1, -1)]
['output1']
[<DataType.HALF: 1>]
[(-1, 128, -1, -1)]
```
How should I test the model with perf test ? I compiled the engine externally and the model is just 2 convolutions for testing with following min max shapes:
```
     --minShapes="input1:2x8x128x128" 
     --optShapes="input1:2x8x256x256" 
     --maxShapes="input1:4x8x512x512"
```
Perf test alway sets the dimensions to `(1,8,1,1)`

Reply:
Good catch.

ORT interprets "-1" of dimension as empty so you can't see the effect of override free dimensions from onnxruntime_perf_test.
ORT uses symbolic dimension for dynamic dimension, so i modified the script to generate dynamic dimension with `free_dim_0` ,` free_dim_1` ...

Also, please use `-f` to specify a free dimension by name not `-F` . So please run like this:
`onnxruntime_perf_test -e tensorrt -M -I -F "free_dim_0:2" -F "free_dim_1:256" -F "free_dim_2:256" trt_path_embed.onnx`

Reply:
Yeah I usually use `-f ` I just tried both for TRT. Using actual names sounds much more reasonable.